core/src/smartcontracts/wasm.rs

```diff
@@ -985,33 +1119,49 @@ trait ExecuteOperationsAsExecutorMut<S> {}
 impl<R, S> import_traits::ExecuteOperations<S> for R
 where
     R: ExecuteOperationsAsExecutorMut<S>,
-    S: state::Wsv + state::WsvMut + state::Authority,
+    S: state::Wsv + state::WsvMut + state::Authority + state::AddExecutedQuery,
 {
     #[codec::wrap]
     fn execute_query(
         query_request: SmartContractQueryRequest,
-        state: &S,
+        state: &mut S,
     ) -> Result<BatchedResponse<Value>, ValidationFail> {
         debug!(%query_request, "Executing as executor");
 
-        let wsv = state.wsv();
-
         match query_request.0 {
             QueryRequest::Query(QueryWithParameters {
                 query,
                 sorting,
                 pagination,
             }) => {
-                let output = query.execute(wsv)?;
-
-                wsv.query_handle().handle_query_output(
-                    output,
-                    NonZeroUsize::new(30_000).expect("30 000 is not zero"),
-                    &sorting,
-                    pagination,
-                )
+                let batched = {
+                    let wsv = state.wsv();
+                    let output = query.execute(wsv)?;
+
+                    wsv.query_handle().handle_query_output(
+                        output,
+                        NonZeroUsize::new(30_000).expect("30 000 is not zero"),
+                        &sorting,
+                        pagination,
+                    )
+                }?;
+                match &batched {
+                    BatchedResponse::V1(batched) => {
+                        if let Some(query_id) = &batched.cursor.query_id {
+                            state.add_executed_query(query_id.clone());
+                        }
+                    }
+                }
+                Ok(batched)
+            }
+            QueryRequest::Cursor(cursor) => {
+                // In a normal situation we already have this `query_id` stored,
+                // so that's a protection from malicious smart contract
+                if let Some(query_id) = &cursor.query_id {
+                    state.add_executed_query(query_id.clone());
+                }
+                state.wsv().query_handle().handle_query_cursor(cursor)
```

Comment:
this looks so duplicated in 3 places

Reply:
Agree, will try to do something with this